### PR TITLE
BAL kernel module - support for RC663 and PN512

### PR DESCRIPTION
**Pull request by PavolCelko**
Christian, please do a review of the code.
I tested the module with the nine Examples and it worked.

These is an issue with the baldev_ioctl. It does not accept 2 as second (cmd) argument. Can you look at it?

Another issue is, I do not know how to #include symbols  like PHBAL_REG_CONFIG_HAL_HW_TYPE from phbalReg.h to the BAL sources. So for now there are hardcoded numbers in the baldev_ioctl().

Pavol

### CODE REVIEW

File: bal/Makefile [hunk 1]
```diff
@@ -4,3 +4,4 @@
 
 obj-m					+= bal.o
 ccflags-y := -I$(src)/include
+EXTRA_CFLAGS += -Wno-error=date-time
```
Comment (christianeisendle) [inline bal/Makefile:7]: What is this used for?

File: bal/bal.c [hunk 2]
```diff
@@ -29,14 +29,21 @@
 #include <linux/spi/bal_spi.h>
 #include <linux/uaccess.h>
 
+//#include <phbalReg.h>
```
Comment (christianeisendle) [inline bal/bal.c:32]: Remove this

File: bal/bal.c [hunk 3]
```diff
@@ -198,7 +309,8 @@ static int bal_spi_remove(struct spi_device *spi)
 
 static int bal_spi_probe(struct spi_device *spi)
 {
-	dev_info(&spi->dev, "Probing BAL driver\n");
+	dev_info(&spi->dev, "Probing BAL kernel module driver\n");
+        printk(KERN_ERR "BAL module MY printk %s\n", __TIME__);
```
Comment (christianeisendle) [inline bal/bal.c:313]: remove this

File: bal/bal.c [hunk 4]
```diff
@@ -152,18 +234,47 @@ static int baldev_release(struct inode *inode, struct file *filp)
...
-	int status = -EINVAL;
-	if (cmd == BAL_IOC_BUSY_PIN) {
-		if (gpio_is_valid(arg)) {
-			status = gpio_request(arg, "BUSY pin");
-			if (!status) {
-				dev_info(&bal.spi->dev, "Setting BUSY pin to %d\n", (unsigned int)arg);
-				gpio_free(bal.busy_pin);
-				bal.busy_pin = (unsigned int)arg;
-				gpio_direction_input(bal.busy_pin);
+	//printk(KERN_ERR "%s : %d fp=%ld cmd=%ld arg=%lx\n", __FUNCTION__, __LINE__, (unsigned long int)filp, cmd, arg);
```
Comment (christianeisendle) [inline bal/bal.c:237]: remove this

File: bal/bal.c [hunk 5]
```diff
@@ -152,18 +234,47 @@ static int baldev_release(struct inode *inode, struct file *filp)
...
+	printk(KERN_ERR "my ioctl - cmd: %d - arg: %lu\n", cmd, arg);
```
Comment (christianeisendle) [inline bal/bal.c:238]: This is for debugging purpose, so KERN_ERR shouldn't be used.
Moreover, one the device is successfully probed and driver is loaded, printk should also not be used.
Please replace them with
dev_err or dev_info or dev_dbg

Example: dev_dbg(&(bal.spi->dev), "my ioctl - cmd: %d - arg: %lu\n", cmd, arg);

File: bal/bal.c [hunk 6]
```diff
@@ -152,18 +234,47 @@ static int baldev_release(struct inode *inode, struct file *filp)
...
+
+	int status = 0; //-EINVAL;
+
+	switch (cmd) {
+		case BAL_IOC_BUSY_PIN:
+			printk(KERN_ERR "busy pin\n");
```
Comment (christianeisendle) [inline bal/bal.c]: remove

File: bal/bal.c [hunk 7]
```diff
@@ -76,14 +88,48 @@ baldev_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
...
+	{
+
+		status = copy_from_user(bal.buffer, buf, count);
+
+		bal.xfers.tx_buf = bal.buffer;
+		bal.xfers.rx_buf = bal.buffer;
+		bal.xfers.len = count;
+
+		//"Fast" bundled exchange for EMVCo compatibility -- only necessary for PN512/RC663 derivatives
+		if( count < 1 )
```
Comment (christianeisendle) [inline bal/bal.c:108]: What is the purpose for checking this here? it should be checked already at the beginning of the function. I wouldn't even check it.

File: bal/bal.c [hunk 8]
```diff
@@ -76,14 +88,48 @@ baldev_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
...
+			return count;
+
+	    //This kenel module does not need to check read/write flag bit. It is upon caller do decide whether read or write.
+		if(bal.MultiRegRW == 1)
+	    {
+	    	//Perform a "MultiRegRead" exchange
```
Comment (christianeisendle) [inline bal/bal.c:114]: What is the difference between this if block and the else block below? I don't see any

**Comment by christianeisendle** [comment]
I'm not going to merge the pull request for now, as there are several issues for which i directly added comments within the diff. 

However, there are several reasons for not merging it:
- The implementation assumes that ioctl is performed to select the IC (PN5180 vs. PN512 vs. RC663.) If this is not done, the behavior is undefined (no default value). PN5180 has to be default
- In case RC663 or PN512 is used, the driver still requires to define a BUSY GPIO within device tree. However, these ICs don't need it. So to clean this up, the IC type should be specified within device tree and not within ioctl (this is anyhow a static decision. A user will never use both frontends in parallel, or in fact, the driver doesn't anyhow support multiple instances as of now 
- PN512 and RC663 use full duplex SPI. Therefore it doesn't make any sense to implement a write function for them but rather just implement read() only. The caller can then anyhow decide if the returned value should be used or ignored. PN5180 supports half duplex SPI only, so here separation in write and read make sense.
- Last not least: I don't think that the multireg option is working. Even if it's working i'm not sure if we pass EMVCo L1 with that, as I fear that the latency in Linux is too high. @michaelneurohr we need to run EMVCo L1 tests on @PavolCelko 's implementation. I also don't want to pollute this driver with IC specific functionality which should actually be done in the HAL. So let's discuss this first.

File: bal/Makefile (hunk 1 above)
Comment (PavolCelko) [inline bal/Makefile:7]: @This line makes that usage of the **TIME** macro returns warning in commpilation istead error.
The **TIME** was used for debugging, to be sure which BAL module was inserted. The line will be removed from the Makefile.

File: bal/bal.c (hunk 7 above)
Comment (PavolCelko) [inline bal/bal.c:108]: Yes, you are righ. Checking on negative value is done by caller. I took it from phbalReg_Stub.c from the another project.

File: bal/bal.c (hunk 8 above)
Comment (PavolCelko) [inline bal/bal.c:114]: There is no difference. I just wanted to keep formal spliting like in LPC project in LPC software package.

**Comment by PavolCelko** [comment]
1,2) I am little confused whether the reader IC is specified within device tree or by ioctl(). Can ioctl change used reader IC?

3) Agree, it should be possible to implement PN512 and RC663 both the read and write operations in read only. The only issue is copying received data at the end, which makes sense for read but for write it can be unnecessary delay. But the copying can be omitted according the write flag (RC663 LSB == 0; PN512 MSB == 0). Are you OK with this "HAL" information within the BAL module?

4) Yes, you are right, Christian. There are some parts in the BAL, which look like HAL dependent. I just copied the solution from phbalReg_Stub.c of the MCU LPC1769 project package. For example there is Multi reg write done by sending 2 bytes in one transfer. In addition the user sapce implementation phbalReg_Stub.c by Donatien also writes multiple registers by 2 bytes in loop. Actually, I tested EMVCo7 example like user with write multi reg implementation just like write reg (no writing by 2 bytes) and works OK on RC663. So I do not understand the splitting sending read and write to multireg and non multireg.
If the BAL implementation with no recognition between multi and single reg read/write passes tests on your site, then the software MultiReg flag variable is redundant. Information HalType about which reader IC is attached is necessary due to PN5180 which has wainitg on busy pin and different read and write approach. So there is no need to recoginse in BAL between RC663 and PN512, except the different position of the read/write flag mentioned in 3) above. To sum it up, we need to recognise between RC663 and PN512 only if we want to skip copying response from reader IC when data written.

**Pull request closed without merging**